Määrien toteumat - Äkilliset ja lisätyöt

jlansineva opened the pull request.

rckopteri committed:
Merge branch 'develop' into VHAR-2266-Lomake-maarien-toteututmille

rckopteri committed:
Merge branch 'develop' into VHAR-2266-Lomake-maarien-toteututmille

jlansineva committed:
lomakkeen generointi osalle komponentteja

jlansineva committed:
lomakkeen komponenttien uudelleeen muotoilua

jlansineva committed:
Merge branch 'develop' into akilliset-hoitotyot

rckopteri committed:
Määrien toteutumien lomake, haku ja kantaan tallennus

rckopteri committed:
Merge branch 'develop' into VHAR-2266-Lomake-maarien-toteututmille

# Conflicts:
#	src/cljs/harja/ui/kentat.cljs
#	tietokanta/src/main/resources/db/migration/R__Toimenpidekoodit_tehtavaryhmat.sql

rckopteri committed:
Merge branch 'VHAR-2266-Lomake-maarien-toteututmille' into toteumat-lisatyot-akilliset-main

# Conflicts:
#	src/cljs/harja/tiedot/urakka/urakka.cljs

rckopteri committed:
Yksitttätisen toteuman hakua parannettu

rckopteri committed:
Nimetään migraatio tiedosto uusiksi

rckopteri committed:
Lisätty urakan tehtävien määrille luotu timestamp, jotta niitä voidaan hakea toteutumat listaukseen ja yksikkötesteihin

rckopteri committed:
Lisää happycase testit toteuman lisäämiselle ja muokkaamiselle.

rckopteri committed:
Merge branch 'develop' into toteumat-lisatyot-akilliset-yksikkotestit

jlansineva committed:
lisätty disablointimahdollisuus tierekisteriosoitteelle - akillisten hoitotöiden front UI

rckopteri committed:
Korjaa määrien toteumien tietokantahakuja

rckopteri committed:
Tee palvelut ja tietokantahaut akillisille hoitotöille

rckopteri committed:
Happycase yksikköttestit

Määrien toteutumille ja äkillisille hoitotöille on yksikkötestit lisäämiselle ja muokkaamiselle

rckopteri committed:
Merge branch 'develop' into toteumat-lisatyot-akilliset-yksikkotestit

rckopteri committed:
Merge branch 'toteumat-lisatyot-akilliset-yksikkotestit' into toteumat-lisatyot-akilliset-main

jlansineva committed:
postausta

jlansineva committed:
Merge branch 'toteumat-lisatyot-akilliset-main' into akilliset-UI

rckopteri committed:
Parannettu määrien toteuman yksikkötestejä

Lisätty muokkauksen tarkistuksia sekä virheellisen toteuman simppeli testi

rckopteri committed:
Merge branch 'develop' into toteumat-lisatyot-akilliset-yksikkotestit

rckopteri committed:
Merge branch 'toteumat-lisatyot-akilliset-yksikkotestit' into toteumat-lisatyot-akilliset-main

jlansineva committed:
wip

jlansineva committed:
Merge branch 'toteumat-lisatyot-akilliset-main' into akilliset-UI

jlansineva committed:
lomakkeen muokkaus tyypin vaihdon yhteydessä

jlansineva committed:
useamman toteuman lisääminen mahdollista

jlansineva committed:
useampi toteuma vain määrämitattaville, lisätty rajaus

jlansineva committed:
lomakkeen munklaamista useampien toteumien kanssa, sijaintitiedot oikeaan paikkaan

jlansineva committed:
datan lähettäminen lomakkeelta bäkkiin

jlansineva committed:
bäkille lähtevän tietorakenteen speksausta

jlansineva committed:
syöttölomake kiinni oikeaan osioon

jlansineva committed:
tehtävien haku syöttölomakkeelle

jlansineva committed:
tehtävien haku äkillisille toimenpiteille ja oikeiden tehtävien filtteröinti

jlansineva committed:
mergetty develop

jlansineva committed:
toteumia voi kirjata eri tyypeille

jlansineva committed:
doseq -> doall for, jotta palautuu id:t ilman atomin tms käpistelyjä

jlansineva committed:
lisätöiden haku ja kirjaus onnistuu

jlansineva committed:
lisätöille migraatiohin dummy-tehtävien luonti

jlansineva committed:
piilotetaan lisätöiden tehtävät sieltä, missä niitä ei tarvita

jlansineva committed:
toteuman muokkaus, sijainnin nappaus ja lomakkeelle mahdollisuus antaa nimi vektorina

jlansineva committed:
toteuman muokkaus mahdollista

jlansineva committed:
tehtävien haun ja filtteröinnin korjaus / päivitys lisätöiden osalta

jlansineva committed:
toteumien poisto on mahdollista

jlansineva committed:
toimenpidekoodit yms fiksed ja koodin yhdistämistä, filtterit listanäkymään

jlansineva committed:
tyyppien laajentaminen listanäkymässä

jlansineva committed:
tyylejä käyttöön paremmin ja sitten sorttausta tehtäville listassa

jlansineva committed:
lomakkeeseen header, ja sitä käyttäen takaisin nappi toteuma-lomakkeelle

jlansineva committed:
tabletyylittelyjä

jlansineva committed:
tabletyylittelyjä

jlansineva committed:
valinta-komponentille vayla-tyyli-vipu

jlansineva committed:
modifierit flex-rowille

jlansineva committed:
tyylivääntöjä ja speksin mukaan viilausta, validointeja

rckopteri committed:
Merge branch 'develop' into akilliset-UI

jlansineva committed:
validointeja ja virheitä näkyviin eri kentille

jlansineva committed:
validointien korjauksia, ja virheitä näkyviin

jlansineva committed:
fiksi, jos ollaan syöttämässä useampaa määrämitattavaa tehtävää ja vaihdetaan tyyppiä, lomake hajoilee. ei hajoile enää

rckopteri committed:
Mahdollista toteuman tehtävien poisto, kun toteuma poistetaan

rckopteri committed:
Korjaa toteutumien haku filtteröimällä poistetut toteuman tehtävät pois hausta

rckopteri committed:
Lisää toteutumien määrien hakuun sijaintitiedot

rckopteri committed:
Lisää toteutumien poiston yhteyteen toteuman tehtävän poisto

rckopteri committed:
Korjaa määrien toteumien ryhmittely, kun käytetään default filttereitä

rckopteri committed:
Korjaa js varoitukset

rckopteri committed:
Toteuma määrien testit päivitetty vastaamaan uudempaa tietomallia

rckopteri committed:
Merge branch 'akilliset-UI' of https://github.com/finnishtransportagency/harja into akilliset-UI

rckopteri committed:
Merge branch 'develop' into akilliset-UI

rckopteri committed:
Toteumat määrien listaussivulle ui tuunauksia.

rckopteri committed:
Korjaa määrien toteuman toimenpide alasvetovalikon rivien korkeus

rckopteri committed:
Estä toteuma määrän lisääminen muille kuin halutuille tehtäville

rckopteri committed:
Korjaa toteumamäärien default tiedot

Sekä listauksesta välitettävän toimenpiteen valinta lomakkeelle

rckopteri committed:
Päivitä määrien toteumien validointeja

rckopteri committed:
Mahdollista radio buttonien disablointi

rckopteri committed:
Päivitä poistomodaalin tiedot

rckopteri committed:
Poista turhat reagent key varoitukset lisäämällä uniikit avaimet

rckopteri committed:
Formatoi koodeja

rckopteri committed:
Korjaa väylä ui:n alasvetovalikon listaitemin korkeudet

rckopteri committed:
Paranna sql lauseiden kommentteja

rckopteri committed:
Korjattu määrien toteuman lomakkeen toiminta

Tehtävien valinta korjattu, toimenpiteen valinta korjattu. Ja pari muuta koodin formatointia

rckopteri committed:
Lisätty poistaminen määrien toteumille

rckopteri committed:
Formatoi prosentit kahden desimaalin tarkkuuteen

rckopteri committed:
Pakotetaan lisätöille ja äkillisille hoitotöille vain yksi mahdollinen toimenpide

rckopteri committed:
Merge branch 'develop' into akilliset-UI

# Conflicts:
#	src/clj/harja/palvelin/palvelut/toteumat.clj

rckopteri committed:
Pakota toimenpide lisätöille ja äkillisille hoitotöille

rckopteri committed:
Formatoi tiedosto

rckopteri committed:
Järjestä toteumat päivämäärän mukaan

Ja korjaa duplikaatit pois hausta

rckopteri committed:
Yksinkertaisat logitusta

rckopteri committed:
Alusta uuden toteuman luonti toimivammin

rckopteri committed:
Näytä toteumat listauksessa avattava rivi vain, kun sen voi oikeasti avata

rckopteri committed:
Alustavasti reittien lisääminen toteumille

rckopteri committed:
Merge branch 'develop' into akilliset-UI

# Conflicts:
#	src/cljs/harja/tiedot/urakka/urakka.cljs

rckopteri committed:
Korjataan uuden tyhjän toteuman lisääminen

Ja korjataan decimaalinen näyttäminen javascriptillä

rckopteri committed:
UI korjauksia

Pakota lomakkeella deskarissa vierekkäin näkyvät elementit mobiilissa alekain.
Laita alasvetovalikon alasnuoli siniseksi.
Poista ylimääräiset marginit lomakkeen isojen koontielementtien sivulta

rckopteri committed:
Lisää yksikkö numerokentälle

rckopteri committed:
Mahdollista vaylä-tyyli sijainti input elementeille

rckopteri committed:
Lisää yksikkö määrien toteumien numerokentälle.

rckopteri committed:
Lisää väylä? tyyppiselle alasvetovalikolle merkintä, kun arvo on valittu

rckopteri committed:
Merge branch 'develop' into akilliset-UI

# Conflicts:
#	tietokanta/testidata/tehtavamaarat.sql

rckopteri committed:
Merge branch 'develop' into akilliset-UI

rckopteri committed:
Korjaa yksikkötestit

rckopteri committed:
Määrien toteumien ui tuunausta ja yksikkötestien korjaus

rckopteri committed:
Refactoroi ja poista turhia kommentteja

rckopteri committed:
Varmista, että urakka löytyy ennen testiä.

Urakka oli ennen atomissa, jota ei oltu keretty täyttää ja testit failasi

rckopteri committed:
Lisää kommentteja, jotta fronttitestit saataisiin toimimaan paremmin

rckopteri committed:
Korjaa tr-kentät testit lisäämällä tarvittavat luokat html:ään

rckopteri committed:
Grid fronttitestejä päivitty tulemaan toimeen uusimman :numero kentän kanssa

rckopteri committed:
Poista cljs-react-test project.clj tiedostosta

rckopteri committed:
Lisätty välistyksiä väylä? tyyppiseen lomakkeeseen

rckopteri committed:
Merge branch 'develop' into akilliset-UI

rckopteri committed:
Siivoa app-state kun uutta määrien toteumaa lisätään

Korjaa sisäinen-label muilta kuin väylä? elementeiltä

rckopteri committed:
Laajenna alasvetovalikon itemin valintaaluetta

rckopteri committed:
Siivoa sijainnit pois app-statesta (ja kannasta) jos ei-sijainttia? checkbox on painettu

rckopteri committed:
Korjaa määrien toteumien lomakkeen default arvojen asetus

rckopteri committed:
Määrien toteumien lomakkeen ensimmäinen validointi poistettu

rckopteri committed:
Viimeksi-muokattu-kenttä lisätty lomakkeen siivottaviin kenttiin

rckopteri committed:
Merge branch 'develop' into akilliset-UI

# Conflicts:
#	tietokanta/testidata/toteumat.sql

rckopteri committed:
Korjataan alasvetovalikon avaus chevronista klikkaamalla mahdollistamalla klikattu-ulkopuolelle funkkarin toiminta

rckopteri committed:
Muuta vähän focuksen näyttämistä väylä-tyyleissä

rckopteri committed:
Hae toteuman id käsin, koska partitioinnin jälkeen sitä ei enää insert palauta

rckopteri committed:
Merge branch 'VHAR-2846' into akilliset-UI

rckopteri committed:
Korjaa määrien toteumien ui:n palstoitus useampaa toteumaa lisättäessä

rckopteri committed:
Päivitä migraatiotiedoston nimi

rckopteri committed:
Muokkaa määrien toteumien tietokantaan tallennus vastaamaan uutta parititioitua taulua

rckopteri committed:
Merge branch 'develop' into akilliset-UI

# Conflicts:
#	src/clj/harja/kyselyt/tehtavamaarat.sql
#	src/cljs/harja/ui/yleiset.cljs
#	tietokanta/src/main/resources/db/migration/V1_833__.sql

rckopteri committed:
Merge branch 'develop' into akilliset-UI

rckopteri committed:
Päivitä sql tiedoston nimi

rckopteri committed:
Korjaa sql tiedostojen järjestys

rckopteri committed:
Korjataan alasvetovalikon toiminta, kun samalla sivulla on useita alasvetovalikoita

rckopteri committed:
Merge branch 'develop' into akilliset-UI

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_834__.sql
#	tietokanta/src/main/resources/db/migration/V1_835__.sql

rckopteri committed:
Muutetaan määrien toteumat käyttämään yksikön sijasta suunnitteluyksikköä

rckopteri committed:
Siivoa useamman määrien toteuman onnistuneen lisäämisen jälkeen lomakkeen tila

rckopteri committed:
Siirrä yksikön labelia hieman vasemmalle

rckopteri committed:
Refaktoroi määrien toteumat ui koodia helpommaksi

rckopteri committed:
Korjaa yksikkötestit

rckopteri committed:
Merge branch 'develop' into akilliset-UI

rckopteri committed:
Ja vielä yksi unohtunut testi laitetaan kuntoon

rckopteri committed:
Merge branch 'develop' into akilliset-UI

Joonas-M-S commented on project.clj:
```diff
@@ -81,6 +81,7 @@
                  [cljs-ajax "0.8.0"]
 
                  [reagent "0.9.1"]
+                 [data-frisk-reagent "0.4.5"]
```
Tämä täällä toista kertaa. Tuo kannattas varmaan siirtää muutenkin `profiles.clj` tiedoston `:dev` osioon, niin tuota ei pakata sitten tuotantoon mukaan.

rckopteri replied:
Kiitos. Joo. Poistin kokonaan. Voidaan tästä jutella joskus erikseen.

Joonas-M-S commented on src/clj/harja/kyselyt/tehtavamaarat.sql:
```diff
@@ -39,7 +39,7 @@ FROM tehtavaryhma tr1
          JOIN toimenpidekoodi tpk3 ON tpk4.emo = tpk3.id
          JOIN toimenpideinstanssi tpi on tpi.toimenpide = tpk3.id and tpi.urakka = :urakka
-WHERE tr1.emo is null
+WHERE tr1.emo is null and tr3.nimi not like '%Lisätyöt%'
```
Nimien käyttö näissä ehdoissa on vähän riskaabelia, kun tuon `tr3.nimi` kentän arvo melko varmasti muuttuu jossain väli. Tuossa `tehtavaryhma` taulukkossa on kenttä `yksiloiva_tunniste`, johon voisit generoida UUID:n, jotta se arvo on sama kaikilla serverillä. Sitten voisit käyttää sitä arvoa tässä ehtona. Jos ymmärsin tämän oikein, niin sen sijaan, että tuo on `tr3.nimi not like '%Lisätyöt%'`, niin se voisi olla `tr3.emo != (SELECT id FROM tr2 WHERE yksiloiva_tunniste=[ UUID TUNNISTE ])`, missä tuo `tr2`:sen UUID on 'Välitaso Lisätyöt' tehtäväryhmän UUID. Tosin ton subqueryn voi varmaan laittaa `WITH` klaussiin, jotta se ajetaan vaan kerran.

rckopteri replied:
Muutettu.

Joonas-M-S commented on src/clj/harja/kyselyt/tehtavamaarat.sql:
```diff
@@ -72,7 +72,7 @@ FROM tehtavaryhma tr1
                    ON tr3.id = tpk4.tehtavaryhma and tpk4.taso = 4 AND tpk4.ensisijainen is true AND
                       tpk4.poistettu is not true AND tpk4.piilota is not true
-         JOIN toimenpidekoodi tpk3 ON tpk4.emo = tpk3.id
+         JOIN toimenpidekoodi tpk3 ON tpk4.emo = tpk3.id and tpk4.nimi not like '%Lisätyö%'
```
Tässä sama kuin ylempänä

rckopteri replied:
Tämä on mergetty developista, mutta näkyy jotenkin tähän branchiin tehtynä. Jätän tämän nyt korjaamatta.

Joonas-M-S commented on src/clj/harja/kyselyt/tehtavamaarat.sql:
```diff
@@ -111,13 +111,16 @@ FROM tehtavaryhma tr1
                    ON tr3.id = tpk4.tehtavaryhma AND tpk4.taso = 4 AND tpk4.ensisijainen is true AND
                       tpk4.poistettu is not true AND tpk4.piilota is not true AND tpk4.nimi NOT IN
                                                                                   (
```
Tässäkin varmaan voi käyttää UUID arvoja

rckopteri replied:
Tässä sama kuin yllä ja sitä ylempänä.

Joonas-M-S commented on src/clj/harja/kyselyt/toteumat.sql:
```diff
@@ -411,6 +416,210 @@ FROM erilliskustannus
+                           toimenpidekoodi tk,
+                           toteuma t,
+                           tehtavaryhma tr1
+                               JOIN tehtavaryhma tr2 ON tr2.id = tr1.emo
```
Toimiiko tämä ihan niinku pitäs? `tr1`:llähän ei pitäs olla emoa, joten tuo `tr2.id = tr1.emo` pitäisi aina olla epätosi. 
Tuossa vissiin tarkoitit tehdä `tr2.emo=tr1.id`. Lisäksi tuossa alempana `WHERE` klaussissa ei ole määritetty `tr1.emo IS NULL`, jolloinka `tr1` sisältää muitakin tehtavaryhmiä kuin vain ensimmäisen tason.

rckopteri replied:
Tämä oliki mielenkiintoinen. Hyvä kun löysit. Noita tr2 ja tr3 ryhmiä ei käytetä mihinkään. Joten se ei haitannut, että tuo join oli väärin. Hieno bongaus! Nyt korjattu

rckopteri replied:
Tämä hoidettu. tr2:sta ei käytetty, niin sen takia haku toimi.

rckopteri replied:
Tutkailin tätä lisää ja alkaa pikkuhiljaa muistamaan, että miten tämä toimi. Tuossa on nuo taulun nimet määritelty väärin päin. eli tr1 pitäisi olla tr3. Eli nimetty ylhäältä alas, mutta haetaan alhaalta ylös.

Joonas-M-S commented on src/clj/harja/kyselyt/toteumat.sql:
```diff
@@ -411,6 +416,210 @@ FROM erilliskustannus
+             LEFT JOIN toteumat t
+                       ON t.toimenpidekoodi_id = ut.tehtava AND t."hoitokauden-alkuvuosi" = ut."hoitokauden-alkuvuosi",
+         toimenpidekoodi tk,
+         tehtavaryhma tr1
```
Täällä sama kuin tuossa `WITH` klaussissa

rckopteri replied:
Jep. tr2 poistettu. Kiitos!

Joonas-M-S commented on src/clj/harja/kyselyt/toteumat.sql:
```diff
@@ -411,6 +416,210 @@ FROM erilliskustannus
+                        AND t.poistettu IS NOT TRUE
+                        AND tt.poistettu IS NOT TRUE
+                        AND t.urakka = :urakka
+                        AND tr1.id = tk.tehtavaryhma
```
Tuo `tr1` taitaa tällä hetkellä tosiaan sisltää kaikki tehtäväryhmät, niin tämä ehto pitänee muuttaa tyylinsä tällaiseen `tk.tehtavaryhma IN (tr1.id, tr2.id, tr3.id)` Tai jos ei tarvi tehdä tuollaista, niin voisiko tuosta `JOIN`:sta ottaa nuo `tr2` ja `tr3` kokonaan pois, kun niitä ei taideta käyttää missään?

rckopteri replied:
Poistettu.

Joonas-M-S commented on src/clj/harja/kyselyt/toteumat.sql:
```diff
@@ -411,6 +416,210 @@ FROM erilliskustannus
+       t.id                       AS toteuma_id,
+       tt.id                      AS toteuma_tehtava_id,
+       t.urakka                   AS urakka,
+       EXTRACT(YEAR FROM alkanut) AS "hoitokauden-alkuvuosi",
```
`alkanut` on vissiin `toteuma` taulukosta. Tässä ei ole tehty smanlaista `CASE` hommaa kuin `WITH` klaussissa.

rckopteri replied:
Hyvä löytö. Case otettu käyttöön.

Joonas-M-S commented on src/clj/harja/kyselyt/toteumat.sql:
```diff
@@ -411,6 +416,210 @@ FROM erilliskustannus
+UNION
+-- ei union all, koska ei haluta duplikaatteja
+-- Haetaan ne tehtävät, joilla ei ole suunniteltua määrää olemassa
+SELECT tk.id                      AS id,
```
Onko tarkoituksella eri taulukon `id`, kuin unionin ensimmäisessä kohdassa?

rckopteri replied:
Kyllä. Rivit haetaan eri tauluista. UNIONissa pitää olla samat kentät. Id:tä ei oikeastaan käytetä mihinkään tässä. Lähinnä otin mukaan, jotta sillä voitaisiin reagentille tarvittaessa eritellä rivit.

Joonas-M-S commented on src/clj/harja/kyselyt/toteumat.sql:
```diff
@@ -411,6 +416,210 @@ FROM erilliskustannus
 WHERE urakka = :urakka
       AND pvm >= :alkupvm AND pvm <= :loppupvm AND poistettu IS NOT TRUE;
 
+-- name: listaa-urakan-maarien-toteumat
```
Tätä queryä vois varmaan `EXPLAIN ANALYZE`:n avulla yrittää vähän optimoida, koska härpitään `toteuma` taulukkoa muutamassa kohdassa eikä `alkanut` kenttä välttämättä ole edes `WHERE` klaussissa käytössä, niin partitioinnista ei välttämättä ole mitään apua. Varmaan @jarnovayrynen sais apuja tuohon hommaan

Joonas-M-S commented on src/clj/harja/kyselyt/toteumat.sql:
```diff
@@ -411,6 +416,210 @@ FROM erilliskustannus
+         toimenpidekoodi tk,
+         toteuma t,
+         tehtavaryhma tr1
+             JOIN tehtavaryhma tr2 ON tr2.id = tr1.emo
```
`tr1.emo` homma ja `WHERE` klaussissa ei olla filtteröity `tr1`:lle kuuluvia kenttiä

rckopteri replied:
Mä en nyt saa tästä kiinni. tr1.id = tk.tehtavaryhma. Eli tehtäväryhmä taulu linkitetään toimenpidekoodiin id:n perusteella.

Joonas-M-S replied:
Tarkoitin tässä sitä, että jos tuon `tr1` on tarkoitus olla se ylin mahdollinen taso, niin tuo `tr1.emo`:n pitäisi olla aina `NULL`, jolloinka tuolla `WHERE` klaussissa pitäs olla `tr1.emo IS NULL` ja tuo `tr2.id = tr1.emo` `JOIN` pitäs varmaan olla toisin päin. Mutta tässä taisikin olla niin, että tuo `tr1` on kolmos taso.

Joonas-M-S commented on src/clj/harja/kyselyt/toteumat.sql:
```diff
@@ -411,6 +416,210 @@ FROM erilliskustannus
+-- name: hae-akillinen-toteuma
+-- Hae yksittäinen äkillinen hoitottyö toteuma muokkaukseen
+SELECT t.id        AS toteuma_id,
+       EXTRACT(YEAR FROM alkanut) AS "hoitokauden-alkuvuosi",
```
Pitäskö tässä käyttää sitä `CASE` hommaa myös?

rckopteri replied:
Lisätty

Joonas-M-S commented on src/clj/harja/kyselyt/toteumat.sql:
```diff
@@ -411,6 +416,210 @@ FROM erilliskustannus
+       tk.suunnitteluyksikko AS yksikko
+    FROM toimenpidekoodi tk,
+         tehtavaryhma tr1
+             JOIN tehtavaryhma tr2 ON tr2.id = tr1.emo
```
`tr1` emo homma myös ja `WHERE` claussissa sen filtteröinti sekä nimien käyttö UUID:n sijasta.

Joonas-M-S commented on src/clj/harja/kyselyt/toteumat.sql:
```diff
@@ -411,6 +416,210 @@ FROM erilliskustannus
+        tk.suunnitteluyksikko AS yksikko
+    FROM toimenpidekoodi tk,
+         tehtavaryhma tr
+    WHERE (tr.nimi like '%Äkilliset hoitotyöt%'
```
UUID

Joonas-M-S commented on src/clj/harja/kyselyt/toteumat.sql:
```diff
@@ -411,6 +416,210 @@ FROM erilliskustannus
+        tk.suunnitteluyksikko AS yksikko
+FROM toimenpidekoodi tk,
+     tehtavaryhma tr
+WHERE tk.nimi like '%Lisätyö%'
```
UUID

Joonas-M-S commented on src/clj/harja/palvelin/palvelut/toteumat.clj:
```diff
@@ -364,6 +369,192 @@
+
+(defn poista-maarien-toteuma! [db user {:keys [urakka-id toteuma-id]}]
+  (if (oikeudet/voi-lukea? oikeudet/urakat-toteumat-kokonaishintaisettyot urakka-id user)
+    (do
```
Poistot transaktion sisälle, kun tätä funktiota kutsutaan myös `:poista-toteuma` palvelusta.

Joonas-M-S commented on src/clj/harja/palvelin/palvelut/toteumat.clj:
```diff
@@ -364,6 +369,192 @@
+          ;; ja "akillinen-hoitotyo" Äkillisille hoitotöille
+          tyyppi (if (nil? tyyppi)
+                   "kokonaishintainen"
+                   (get {:maaramitattava "kokonaishintainen"
```
Olisiko `case`:n käyttö tässä idiomaattisempaa Clojurea?

Joonas-M-S commented on src/clj/harja/palvelin/palvelut/toteumat.clj:
```diff
@@ -364,6 +369,192 @@
+        (throw+ {:type "Error"
+                 :virheet [{:koodi "ERROR" :viesti "Tarkista määrä."}]})
+        (jdbc/with-db-transaction [db db]
+                                  (doall
```
`doall` ja `for` yhdistelmän voi korvata `doseq`:lla, jos ei olla kiinnostuneita paluuarvosta

Joonas-M-S commented on src/clj/harja/palvelin/palvelut/toteumat.clj:
```diff
@@ -364,6 +369,192 @@
+                                           toteuma-tehtava-id :toteuma-tehtava-id
+                                           poistettu :poistettu
+                                           {:keys [numero alkuosa alkuetaisyys loppuosa loppuetaisyys]} :sijainti
+                                           :as _toteuma} toteumat]
```
Destrukturoinnin saa pienempään tilaan käyttämällä tuota `:keys` ylätasollakin
```
{:keys [maara lisatieto tehtava toteuma-id toteuma-tehtava-id poistettu]
 {:keys [numero alkuosa alkuetaisyys loppuosa loppuetaisyys]} :sijainti
 :as _toteuma}
```

Ohan tuo kyllä hyvinkin lukija kohtasta, että kumpi on "parempi", eli selkeämpi lukusta.

Joonas-M-S commented on src/cljs/harja/tiedot/urakka/toteumat/maarien_toteumat.cljs:
```diff
@@ -0,0 +1,664 @@
+
+(defn validoinnit
+  ([avain lomake indeksi]
+   (let []
```
`let` jääny

Joonas-M-S commented on src/cljs/harja/tiedot/urakka/toteumat/maarien_toteumat.cljs:
```diff
@@ -0,0 +1,664 @@
+                            [::t/toteumat i ::t/lisatieto] (validoinnit ::t/lisatieto lomake)])
+                         (range (count toteumat))))))
+
+(defn- hae-tehtavat-tyypille
```
Tämä funktio ei vissiinkään tarvitse `app` argumenttia mihinkään, niin pitäisikö vain jättää pois?

Joonas-M-S commented on src/cljs/harja/tiedot/urakka/toteumat/maarien_toteumat.cljs:
```diff
@@ -0,0 +1,664 @@
+                         :paasta-virhe-lapi? true})
+     app)))
+
+(defn- poista-toteuma [id app]
```
Tämäkään ei tarvitse `app` argumenttia

Joonas-M-S commented on src/cljs/harja/tiedot/urakka/toteumat/maarien_toteumat.cljs:
```diff
@@ -0,0 +1,664 @@
+      true
+
+      (contains?
+        (into #{}
```
Jos tuo `filtterit` on mappi boolean arvoja, niin tuota `into {}` hommaa ei tarvi välissä, kun `keys` toimii myös `seq`:lle `MapEntry`:jä.
```
=> (keys (seq {:a 1 :b 2}))
(:a :b)
```

`into #{}` ja `mapcat` sijasta saattaisi olla idiomaattisempaa käyttää `clojure.set/union`:ia. Eli tyyliinsä:
```
(transduce (comp (keep (fn [[tyyppi arvo]]
                         (when (true? arvo)
                           tyyppi)))
                 (map filtteri->tyyppi))
           clojure.set/union
           #{}
           filtterit)
```

Joonas-M-S commented on src/cljs/harja/tiedot/urakka/toteumat/maarien_toteumat.cljs:
```diff
@@ -0,0 +1,664 @@
+        ryhmitelty-tr (group-by :tehtavaryhma
+                                ryhmiteltavat-filtteroitu)]
+    (sort-by first
+             (into {}
```
ei tarvita `into {}` välissä

Joonas-M-S commented on src/cljs/harja/tiedot/urakka/toteumat/maarien_toteumat.cljs:
```diff
@@ -0,0 +1,664 @@
+    t))
+
+(defn- vaihda-toimenpide-tyypin-mukaan [app tyyppi]
+  (let [toimenpide
```
Tämä `let` jäänyt vissiin printtailuista?

Joonas-M-S commented on src/cljs/harja/tiedot/urakka/toteumat/maarien_toteumat.cljs:
```diff
@@ -0,0 +1,664 @@
+                                            toteumat)
+          {:keys [validoi] :as validoinnit} (toteuma-lomakkeen-validoinnit lomake)
+          {:keys [validi? validius]} (validoi validoinnit lomake)
+          toteumat (mapv #(into {}       ; siivotaan namespacet lähetettävästä
```
Meillä on tätä varten tommonen kirjasto kuin https://github.com/Jarzka/namespacefy käytössä. Eli vois käsittääkseni tehdä tän näin:
`(mapv namespacefy/unnamespacefy toteumat)`

Joonas-M-S commented on src/cljs/harja/tiedot/urakka/urakka.cljs:
```diff
@@ -64,37 +72,43 @@
+                                                   (get-in lomake polku)))))))
+                  (update kaikki :validi?
+                          (fn [v?]
+                            (not
```
Olisiko `every?` tässä idiomaattisempi?

Joonas-M-S commented on src/cljs/harja/tiedot/urakka/urakka.cljs:
```diff
@@ -64,37 +72,43 @@
                   :kulut/lisatyon-lisatieto    [ei-nil ei-tyhja]
                   :kulut/toimenpideinstanssi   [ei-nil ei-tyhja]})
 
+(defn validoi!
```
Kiva funktio kyllä. Ja tuo `silloin-kun` ja `ei-pakollinen` myös.

Joonas-M-S commented on src/cljs/harja/ui/lomake.cljs:
```diff
@@ -472,13 +580,15 @@ Ryhmien otsikot lisätään väliin Otsikko record tyyppinä."
                         [:h3.lomake-ryhman-otsikko (:otsikko otsikko)]
                         rivi-ui]
-                       (with-meta rivi-ui {:key i}))))
+                       ^{:key (str "rivi-ui-with-meta-" i)}
```
Tässä ei tarvinne tuota `^{:key ...}` kun tuo `:key` meta annetaan ajon aikana `rivi-ui`:n evaluoidulle arvolle. Muistaakseni tuo `^` reader-macro toimi niin, että sen määrittämä meta annetaan sitä seuraavalle symbolille käännön aikana. Eli nyt sen määrittämä `:key` meta annetaan tolle `(with-meta ...)` listalle.

Joonas-M-S commented on src/cljs/harja/views/urakka/kulut.cljs:
```diff
@@ -542,26 +554,44 @@
+                                        (= "lisatyo" (:maksueratyyppi (first kohdistukset))) false
+                                        :else true)
+                     :disabloitu?     (not= 0 haetaan)
+                     :muutos-fn       #(let [kohdistusten-paivitys-fn (if (.. % -target -checked)
```
Jos määrittää anonyymin funktion näin argumenttina jollekin Reagentin komponentille, niin tuo komponentti (`vayla-radio`) renderöidään joka kerta, kun tämä (`tehtavien-syotto`) komponentti renderöidään vaikkei argumentit olisikaan muuttunut. Tää `muutos-fn` kannattaisi siis määrittää joko tämän (`tehtavien-syotto`) renderöinti funktion ulkopuolella tai sitten wrapata se `reagent.core/partial`:in sisään.

jlansineva replied:
hyvä muistutus, nämä fiksailtu

Joonas-M-S commented on src/cljs/harja/views/urakka/toteumat/maarien_toteuma_lomake.cljs:
```diff
@@ -0,0 +1,395 @@
+    [:span {:style {:padding-right "1rem"}}
+     [napit/yleinen-toissijainen
+      "Peruuta"
+      (fn []
```
`reagent.core/partial`

Joonas-M-S commented on src/cljs/harja/views/urakka/toteumat/maarien_toteuma_lomake.cljs:
```diff
@@ -0,0 +1,395 @@
+                 [:div.row
+                  [napit/poista
+                   "Poista toteuma"
+                   #(modal/nayta! {:otsikko "Haluatko varmasti poistaa toteuman?"}
```
r/partial

Joonas-M-S commented on src/cljs/harja/views/urakka/toteumat/maarien_toteuma_lomake.cljs:
```diff
@@ -0,0 +1,395 @@
+      [napit/takaisin "Takaisin" #(tyhjenna-lomake! nil) {:vayla-tyyli? true :teksti-nappi? true}]]
+     [ui-lomake/lomake
+      {:muokkaa! (fn [data]
+                   (do
```
Tätä `do`:ta ei tarvinne

Joonas-M-S commented on src/cljs/harja/views/urakka/toteumat/maarien_toteumat.cljs:
```diff
@@ -0,0 +1,283 @@
+                             {:otsikko "Kaikki" :id 0}
+                             (:valittu-toimenpide app))
+        valittu-hoitokausi (if (nil? (get-in app [:hoitokauden-alkuvuosi]))
+                             2019
```
Onko `valittu-hoitokausi` aina kaikille urakoille 2019, jos `:hoitokauden-alkuvuosi` ei ole määritetty? Nyt 2020 alkaville urakoille myös?

jlansineva replied:
hyvin bongattu, tämä pääsi lipsahtamaan hardkoodattuna.

Joonas-M-S reviewed (commented):
Nuo refaktorointikommentit voi jättää huomiotta, jos niitä ei jaksa korjata tai on eri mieltä.

Tuolla oli koodissa jonkin verran pois kommentoitua koodia, niin nekin voisi poistaa kokonaan ennen mergeämistä.

jlansineva committed:
hoitokauden alkuvuodesta pois hardkoodattu arvo, ylimääräinen do block pois

jlansineva committed:
fiksattu anonyymejä funkkareita partialien sisään

jlansineva committed:
poistettu kommentoitua koodia ja r/partialisoitu anomuumeja funkkareita

jlansineva committed:
some -> every?

jlansineva committed:
transaktio lisätty poistoon

jlansineva committed:
vaihdettu if caseksi

jlansineva committed:
siistitty destrukturointia

jlansineva committed:
turha let pois

jlansineva committed:
turhat parametrit pois

jlansineva committed:
poistettu ylimääräinen let

jlansineva committed:
namespacefylla siivoillaan namespaceja

rckopteri committed:
Katselmointikommenttien perusteella korjaukset

rckopteri committed:
Merge branch 'develop' into akilliset-UI

rckopteri committed:
Määrien toteumien tietokantahakuja muutettu katselmointikommenttien perusteella

rckopteri committed:
Merge branch 'develop' into akilliset-UI

maaritlaine committed:
Ota voimassaolo huomioon tehtävälistaa hakiessa.

rckopteri committed:
Lisätään uusi partitio toteuma taululle

Lisätään urakka_id toteuma_tehtava taululle ja sille uusi indeksi

rckopteri committed:
Päivitetään testidataan toteuma_tehtava taulun urakka_id

rckopteri committed:
Lisätään api toteutuksiin urakka_id toteuma_tehtava taulun käsittelyihin

rckopteri committed:
Pakotettaan kaikki tuntemattomat toteumatyypit määrämitattavaan lomakkeeseen

rckopteri committed:
Lisätään ajax-loader UI:lle, kun haetaan toteumia serveriltä

rckopteri committed:
Käytetään lyhennettyjä kolumnin nimiä, jotta tiedon siirto serverin ja clientin välillä pienenee

rckopteri committed:
Päivitä totetumien hakua lyhemmäksi ja nopeammaksi

rckopteri committed:
Merge branch 'akilliset-UI' of https://github.com/finnishtransportagency/harja into akilliset-UI

rckopteri committed:
Merge branch 'develop' into akilliset-UI

rckopteri committed:
Merge branch 'develop' into akilliset-UI

# Conflicts:
#	src/cljs/harja/views/urakka/toteumat.cljs

jarnovayrynen commented on tietokanta/src/main/resources/db/migration/V1_839__.sql:
```diff
@@ -0,0 +1,127 @@
+-- Ennen uusien partitioiden synnyttämistä, päivitetään yhden härön toteuman ajankohtaa
+UPDATE toteuma SET alkanut = '1000-01-01 00:00:00' WHERE id = 5606292;
```
Miksi migraatiossa tuotantokantaspesifinen datafiksi? Eikö tällainen kannattaisi tehdä iahn vaan transaktiossa suoraan komentokehoteelta? Saataisiin siistimpi filu.

jarnovayrynen commented on tietokanta/src/main/resources/db/migration/V1_839__.sql:
```diff
@@ -0,0 +1,127 @@
+ALTER TABLE toteuma_250101_991231 ALTER COLUMN poistettu SET NOT NULL;
+-- set default false
+ALTER TABLE toteuma ALTER COLUMN poistettu SET DEFAULT FALSE;
+ALTER TABLE toteuma_010101_191001 ALTER COLUMN poistettu SET DEFAULT FALSE;
```
meniskö NOT NULL ja DEFAULT FALSE samalla komennolla?

rckopteri replied:
Yritin, en osannut laittaa. (normaalisti kun lisätään uutta kolumnia create tablessa, niin onnistuu, mutta en tiedä miksi se aina valitti, kun alterissa koitin)

jarnovayrynen commented on tietokanta/src/main/resources/db/migration/V1_839__.sql:
```diff
@@ -0,0 +1,127 @@
+
+-- Muokataan toteuma_tehtava taulu suorituskyvyltään paremmaksi
+-- Lisätään urakkaindeksiä varten uusi columni
+ALTER TABLE toteuma_tehtava ADD COLUMN urakka_id INTEGER;
```
Tämähän luo redundanssia, kun urakka-id on toteumaviitteen kautta johdettavissa. Kenties kuitenkin ihan OK, jos on todennettavissa suorituskykyhyöty.

Joonas-M-S replied:
Pitäskö tuohon tehdä samalla foreign key constraint  `urakka` taulukkoon? Tosin se constraint pitänee laittaa tonne 841 migraatiofiluun, update hommien jälkeen.

rckopteri replied:
Erittäin hyvä nosto. Jos tuollainen haluttaisiin, niin se pitäisi olla urakka<--->toteuma taulujen välillä (mitä ei ole) ja sitten tähän tauluun se olisi pitänyt laittaa toteuma <---> toteuma_tehtava linkiksi. (jota ei ole) Tehdäänkö tästä erillinen bugi?

jarnovayrynen commented on tietokanta/src/main/resources/db/migration/V1_839__.sql:
```diff
@@ -0,0 +1,127 @@
+-- Muokataan poistettu kenttä sisältämään pelkästään true/false arvoja, koska se parantaa indeksin toimintaa
+UPDATE toteuma_tehtava set poistettu = false where poistettu is null;
+ALTER TABLE toteuma_tehtava ALTER COLUMN poistettu SET NOT NULL;
+ALTER TABLE toteuma_tehtava ALTER COLUMN poistettu SET DEFAULT FALSE;
```
taas: meniskö nää kaks samaan komentoon? vai antaako alter column tehdä vain yhden kerrallaan

jarnovayrynen commented on tietokanta/src/main/resources/db/migration/V1_840__.sql:
```diff
@@ -0,0 +1,21 @@
+
+-- Päivitä toteuma taulujen indeksejä, jotta suorituskyky paranee
+CREATE INDEX CONCURRENTLY toteuma_urakka_alkanut_idx ON toteuma (urakka, alkanut, poistettu);
+CREATE INDEX CONCURRENTLY toteuma_010101_191001_urakka_alkanut_poistettu_idx ON toteuma_010101_191001 (urakka, alkanut, poistettu);
```
Jos luodaan tää indeksi kaikille uusille partitioillekin, pitäisi varmaan päivittää se tänne: 
CREATE OR REPLACE FUNCTION luo_toteumataulun_partitio(alkupvm DATE, loppupvm DATE)
muuten tuo luo_toteumataulun_partitio jää epätilaan, jos jatkossa halutaan käyttää sitä lohkomiseen.

Muutoinkin pelkään vähän että insertit hidastuu kun luomme toteumatauluun vielä lisää indeksejä. Onkohan tässä mahdollista  DROPata ensin tämä indeksi: toteuma_tyyppi_urakka_alkanut_idx
ja sitten luoda tämä uusi ikäänkuin tuon tilalle? Ottaisko yhä kiinni kyselyyn?

rckopteri replied:
Pitäisköhän melkein tuo indeksin poisto testata. Jotta pystyn testaamaan, niin pääsen ajamaan nuo päivitykset singletonille uudestaan. Se vie aina hetken. Samalla voisi tietenkin muutenkin körmyyttää isommallakin köörillä kaikkia toteutumiin liittyviä asioita singletonia vasten.

jarnovayrynen reviewed (approved):
Laitoin joitakin kommentteja. Jos vaan oot hyvin explain analyzellä testannut, varsinkin tuon urakka-id:n lisäämisen toteuma_tehtävä tauluun että se varmasti ottaa siihen kiinni ja tuo hyötyjä, niin näyttäis hyvältä. tuo uus partitioiden indeksin luonti pitäs vielä varmasti uittaa sinne luo_toteumataulun_partitio funkkariin niin pysyy tulevaisuudessakin toimintakykyisenä se funktio

jarnovayrynen commented:
Huomatkaa että mun approve koski nyt vain toteuma ja toteumatehtävä tauluun liittyviä kanta- ja indeksimuutoksia, noita muita juttuja en tunne enkä katsonut lainkaan.

rckopteri committed:
Yksinkertaistettu määrien toteuman hakua pakottamalla ajanjakso haulle

rckopteri committed:
Yksikkötestien korjaus koska ajanjakso (hoitojakso) on pakollinen

rckopteri committed:
Merge branch 'develop' into akilliset-UI

# Conflicts:
#	project.clj
#	tietokanta/src/main/resources/db/migration/V1_837__.sql

rckopteri committed:
Migraatiotiedoston konfliktien takia uudet tiedostojen sisältö laitettu uusiksi

rckopteri committed:
Migraatio tiedostoissa non-transactional ja transactional korjaus

rckopteri committed:
Päivitä yksikkötesteihin oikean tehtävän haku kannasta

rckopteri committed:
Siirrettävän datan optimoinnista johtuva otsikkoiden puuttuminen korjatttu

jarnovayrynen commented:
Testailin läpi, toteumataulun puolesta kaikki keissit toimi enkä löytänyt mitään ihmeellisiä ongelmia suorituskyvystäkään. Mergeä vaan sitten kun nuo muut osa-alueetkin on katselmoitu ja testit vihreänä.

jlansineva committed:
muutettu tehtävän hakua toimenpiteellä määrien toteumissa, jotta muokatessa määrien toteumaa haetaan muokattava tehtävä oikein

rckopteri committed:
Lisää toteuma_materiaalille urakka_id ja uusi indeksi

Ja täydennetään puuttuviin testidatoihin käyttäjän ja urakan tietoja

rckopteri committed:
Ota suolatoteumien materiaalihaussa käyttöön toteuma_materiaalin uudempi indeksi ja nopeuta hakua valtavasti

rckopteri committed:
Ota käyttöön päivitetyt määrien toteumat -tietokantahaut

Nyt mukana on järjestelmästä tulleiden tietojen summaus ja toteuma_materiaalitaulun käyttö

rckopteri committed:
Ota käyttöön uudistuneet määrien toteumat tietokantahaut

rckopteri committed:
Päivitä määrien toteumien ui puoli vastaamaan uudistunutta backendiä

rckopteri committed:
Merge branch 'akilliset-UI' of https://github.com/finnishtransportagency/harja into akilliset-UI

# Conflicts:
#	src/cljs/harja/tiedot/urakka/toteumat/maarien_toteumat.cljs
#	src/cljs/harja/views/urakka/toteumat/maarien_toteumat.cljs

rckopteri committed:
Korjataan tehtävän automatic valinta määrien toteuman lomakkeeseen

rckopteri committed:
Merge branch 'develop' into akilliset-UI

# Conflicts:
#	src/clj/harja/kyselyt/tehtavamaarat.sql
#	tietokanta/src/main/resources/db/migration/V1_838__.sql

rckopteri committed:
Concurrently luotua indeksiä ei voi flywayssa sisällyttää samaan tiedostoon non-concurrently

ajojen kanssa. Joten räjäytetään samat tietokantalauseet useampaan tiedostoon.

rckopteri committed:
Merge branch 'develop' into akilliset-UI

rckopteri committed:
Määrien toteumissa varoitusmerkin näyttö, jos suunnitelma ja toteuma eivät täsmää

rckopteri committed:
päivitä toteuma_materiaalit tauluun urakka_id myös testiaineistoihin

rckopteri committed:
Määrien toteumien testeihin päivitys parametrien nimiin.

rckopteri committed:
Ja vielä yksi puuttunut testiparametrimuutos

rckopteri committed:
Yritä korjata migraatiotiedostojen konflikti

rckopteri committed:
Merge branch 'develop' into akilliset-UI

rckopteri committed:
Vielä yksi spinneri helpottamaan käyttäjiä odottelussa, kun määrien toteumat ei ole ihan millisekuntibusinesta.

rckopteri committed:
Merge branch 'develop' into akilliset-UI
